build: Add pgm to build dependent repos [skip ci]

karl-lunarg opened the pull request:
New Python script to build dependent repos using "known-good" commits.

mikes-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+    repos = GetGoodRepos(args)
+
+    print('Starting builds in {d}'.format(d=abs_top_dir))
+    for repo in sorted(repos, key=lambda x: x.order):
```
order seems redundant with json file order, or better, dependency order

you _could_ do a quick and dirty bubble sort for dependencies, but I admit it isn't super readable:

    for i in range(len(repos)):
        for j in range(i + 1, len(repos)):
            if 'deps' in repos[i] and repos[j]['name'] in [d['reponame'] for d in repos[i]['deps']]:
                repos[i], repos[j] = repos[j], repos[i]

karl-lunarg replied:
I could remove order and they rely on json file order.  TBH, the glslang code sorted by the subrepo name so that nested repos would be cloned in top-to-bottom order, so it was easy for me to keep the notion of sorting, but by a different key.  These repo lists should not be so long that we need a lot of complexity here.

karl-lunarg commented:
> Do we eventually want to put some of the recipes into BUILD.md?

Sorry, I meant and forgot to add a blurb to BUILD.md about that.  I will add it soon.

karl-lunarg commented on scripts/known_good.json:
```diff
@@ -0,0 +1,86 @@
+  "repos" : [
+    {
+      "name" : "glslang",
+      "site" : "github",
```
This was meant to facilitate using the `--site` option to switch between using GitHub and GitLab for cloning the dependent repos.  I'll address this better in your comment starting "I do not understand..."

karl-lunarg commented on scripts/known_good.json:
```diff
@@ -0,0 +1,86 @@
+      "subrepo" : "KhronosGroup/glslang",
+      "subdir" : "glslang",
+      "builddir" : "glslang/build",
+      "installdir" : "glslang/build/install",
```
Yes.  And they have to be relative to the top dir.  I don't think that we want to check this file in if it contains absolute paths, because such paths won't make sense on everyone's system.  "Top dir" is either your current directory or can be specified with `--dir`, which can be relative or absolute.

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+SITE_TO_HOST = {
+    'github': 'https://github.com/',
+    'gitlab': 'git@gitlab.khronos.org:'
+}
```
This was probably put in place for glslang to address their particular use case.  When using the gitlab host, a different file (the one ending in "khr" in the glslang repo) has the gitlab-specific repo names and commits in it.

We could easily add a known_good_khr file that points to the gitlab counterparts of the dependent repos.

I honestly didn't exercise the "alternate" gitlab paths.  How about if we leave this in place and see what is needed from the people who are working with other forks or repos?  I'd be happier if someone who routinely works with the gitlab half would make the khr file and any other needed improvements.

mikes-lunarg replied:
With the old update_external_sources.py system I would update the url and commit ID in the same file on the gitlab fork instead of having a separate file.  That is also why I was asking about the repo url changing in another comment.

Having a second file and passing "--site gitlab" will work but it feels more complicated than it ought to be.

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+        """
+        # Required JSON elements
+        self._json = json
+        self.name = json['name']
```
Thanks, this is good.  I'll try it and see what it looks like.

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+def command_output(cmd, directory, fail_ok=False):
+    """Runs a command in a directory and returns its standard output stream.
+
+    Captures the standard error stream.
```
I think "Captures" means that it is "eating" stderr by sending it to `_` so it doesn't show up anywhere.  But I will check again.

mikes-lunarg replied:
That was my interpretation of "Captures", but you need `stderr=subprocess.PIPE` in the `Popen` constructor for it to work.  The default behavior will cause stderr to be inherited from the parent process.

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+def command_retval(cmd, directory):
+    """Runs a command in a directory and returns its return value.
+
+    Captures the standard error stream.
```
I'll apply the same cure.

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+
+    def HasCommit(self):
+        """Check if the repository contains the known-good commit."""
+        return 0 == subprocess.call(
```
I'll look into using the utility in this and other places.

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+            ret_code = subprocess.call(p.split())
+            if ret_code != 0:
+                sys.exit(ret_code)
+        if platform.system() == 'Linux':
```
Good idea

karl-lunarg commented on scripts/update_deps.py:
```diff
@@ -0,0 +1,630 @@
+        explicitly create branches when doing any meaningful work.
+        """
+        remotes = command_output(['git', 'remote'], self.subdir).splitlines()
+        if b'known-good' not in remotes:
```
Unlikely that the url would change?  This function is probably going away.

tobine commented:
@karl-lunarg can you please beef up the commit message, BUILD.md, and help output for update_deps.py?
Based on the "help" it seems that this will only pull & build the "dependent repos" and not Validation-Layers, correct?
The help could include the dependent repos and a reference to the known-good file that's used to determine repo revisions.
Also, do the options apply to every dependent repo? As in "--clean-build" will delete every build dir for every dependent repo?

BUILD.md should explain the build flow when using update_deps.py script.

This update looks fantastic, by the way, and is a huge step in the right direction.
I know that the code has answers to my questions, but I'm being lazy and just considering what's the user experience if I'm just using the help and docs. The script greatly improves the build experience so let's make sure it's easy to know that it's there and how to use it.

karl-lunarg commented:
@tobine Yes, I plan to add a good description of all of this to the BUILD.md file.

>Based on the "help" it seems that this will only pull & build the "dependent repos" and not Validation-Layers, correct?

Yes.  The assumed workflow is that the user is going to be building the "home" (Validation-Layers) repo with their own options, etc.   They've also already "pulled it" with the revision they want. However, the `update_deps.py` leaves a `helper.cmake` file in the top dir that contains CMake commands for setting things like VULKAN_LOADER_INSTALL_DIR that set all the CMake variables that the home repo depends on.  The user can then `cmake -C helper.cmake` and add their own specific options.

>The help could include the dependent repos and a reference to the known-good file that's used to determine repo revisions.

OK, will look into that

>Also, do the options apply to every dependent repo? As in "--clean-build" will delete every build dir for every dependent repo?

Yes, and then it starts building again.  I need to add info to help that says it does this clean before building.  It doesn't just clean and stop.

> BUILD.md should explain the build flow when using update_deps.py script.

It will.  There's already some text for that in the large comment block at the top of the script, which you can read without needing to read the code.  I'll be placing many of these elements in the BUILD.md file.

karl-lunarg commented:
> Are there plans to incorporate googletest repo into this flow? Note that following the current recommended build flow won't build vk_layer_validation_tests.

This scope of this effort is constrained to updating and building the _dependencies_.  The intent is to relieve the developer of the tedium of gathering the dependent repos at the right versions and building them.

We can't really know what the use case is for each user of this repo.  Some might want to build the tests and others might just want to use the layers.  We also don't know what other options or tweaks someone might want to use on this repo, so we generally leave the configuration of this repo up to the user.  The configuration detail around googletest is well-explained in BUILD.md and is consistent with the glslang model.

So, no googletest in the flow for building _this_ repo.  I purposely left it out of the dependent repo flow as well because the "home" repo does not need the tests in the dependent repos, for most use cases.  

Note that this script, as it stands, should be able to be placed in other repos like Vulkan-Loader and Vulkan-Tools with (hopefully) no modifications.  Only the `known-good.json` file needs to be modified to configure the dependent builds.  Mixing in the taking care of aspects of the "home" repo detracts from this characteristic.

> Also, do you plan a follow-on PR to update .travis.yml to use this script?

Absolutely.   I plan to work on it next/shortly.

karl-lunarg commented:
@mikes-lunarg : I resolved your comments as follows:

- site and subrepo gone; replaced by a url item, like we used to use in the revisions files
- script looks for a single file now; no option for file with a "gitlab" name.   I was thinking you would have wanted separate files for the gitlab shadows.
- the trick to init the `self` dictionary did turn out to be "too magic".  The problem is that the entries are not defined until runtime and that makes the Python linter freak out.
- fixed command_output to swallow stderr.  Side effect is that the console output is a bit nicer.
- added in the mac stuff
- the "add remote" stuff is gone
- removed the order attribute
